module/zfs/dsl_dataset.c

```diff
@@ -1212,6 +1241,26 @@ dsl_dataset_create_sync(dsl_dir_t *pdd, const char *lastname,
 
 	dsl_deleg_set_create_perms(dd, tx, cr);
 
+	/*
+	 * If we are creating a clone and the livelist feature is enabled,
+	 * add the entry DD_FIELD_LIVELIST to ZAP.
+	 */
+	if (origin != NULL &&
+	    spa_feature_is_enabled(dp->dp_spa, SPA_FEATURE_LIVELIST)) {
+		/*
+		 * Livelists are not created for filesystems based on the
+		 * origin snap since they're not actually clones and
+		 * wouldn't benefit from using livelists.
+		 */
+		ASSERT3P(origin, !=, dp->dp_origin_snap);
```

Comment:
Can there be a case where origin != NULL but it is NOT dp->dp_origin_snap?  If not (which the assert would lead me to believe), why is the if condition above not for origin == dp->dp_origin_snap?

Reply:
It took me awhile to figure this out (I'd totally forgotten where that assertion came from in the first place..) but basically it has to do with how the function is invoked and the implication that if we are creating a filesystem (a non-clone), we won't have the origin passed as equal to `dp->dp_origin_snap` (even though it eventually gets set to that). 

I think the point of having there assertion there was to firm up the assumption that anything with a non-NULL origin really is a clone. However, It's kind of complicated and non obvious. 

I've moved it to outside the clone case and added a more general comment that I think is easier to understand.